altair/utils/data.py

```diff
@@ -157,65 +204,115 @@ def sample(
         return None
 
 
-class _JsonFormatDict(TypedDict):
-    type: Literal["json"]
+_FormatType = Literal["csv", "json"]
 
 
-class _CsvFormatDict(TypedDict):
-    type: Literal["csv"]
+class _FormatDict(TypedDict):
+    type: _FormatType
 
 
-class _ToJsonReturnUrlDict(TypedDict):
+class _ToFormatReturnUrlDict(TypedDict):
     url: str
-    format: _JsonFormatDict
+    format: _FormatDict
 
 
-class _ToCsvReturnUrlDict(TypedDict):
-    url: str
-    format: _CsvFormatDict
+@overload
+def to_json(
+    data: Literal[None] = ...,
+    prefix: str = ...,
+    extension: str = ...,
+    filename: str = ...,
+    urlpath: str = ...,
+) -> partial: ...
 
 
-@curried.curry
+@overload
 def to_json(
     data: DataType,
+    prefix: str,
+    extension: str,
+    filename: str,
+    urlpath: str,
+) -> _ToFormatReturnUrlDict: ...
+
+
+def to_json(
+    data: Optional[DataType] = None,
     prefix: str = "altair-data",
     extension: str = "json",
     filename: str = "{prefix}-{hash}.{extension}",
     urlpath: str = "",
-) -> _ToJsonReturnUrlDict:
+) -> Union[partial, _ToFormatReturnUrlDict]:
```

Comment:
Could you elaborate on what the `@overload` decorator does here. I thought its main purpose was to help type checking understand functions that support different combinations of arguments/types, but in this case it seems like there is only one combination of arguments?

Reply:
@joelostblom does the first part of https://github.com/vega/altair/pull/3426#discussion_r1638279005 help with this at all? 

The reason for `@overload` and the pattern used are the same. 

In short, when you call these functions:
- *without* `data` they return a function, [functools.partial](https://docs.python.org/3/library/functools.html#functools.partial)
- *with* `data` they return the type specified by the matching `@overload`

Reply:
Thanks for elaborating, I will look into it more to fully understand it more deeply, but it's not blocking in any way.